engine/worker_task_queue.go

```diff
@@ -0,0 +1,165 @@
+package engine
+
+import (
+	"context"
+	"github.com/xdblab/xdb/common/log"
+	"github.com/xdblab/xdb/common/log/tag"
+	"github.com/xdblab/xdb/config"
+	"github.com/xdblab/xdb/persistence"
+	"math/rand"
+	"time"
+)
+
+type workerTaskQueueSQLImpl struct {
+	shardId int32
+	store   persistence.ProcessStore
+	logger  log.Logger
+	rootCtx context.Context
+	cfg     config.Config
+
+	pollTimer         TimerGate
+	commitTimer       TimerGate
+	processor         WorkerTaskProcessor
+	tasksToCommitChan chan persistence.WorkerTask
+	// the starting sequenceId(inclusive) to read worker tasks
+	currentReadCursor         int64
+	pendingTaskSequenceToPage map[int64]*workerTaskPage
+	completedPages            []*workerTaskPage
+}
+
+type workerTaskPage struct {
+	minTaskSequence int64
+	maxTaskSequence int64
+	pendingCount    int
+}
+
+func NewWorkerTaskProcessorSQLImpl(
+	rootCtx context.Context, shardId int32, cfg config.Config, store persistence.ProcessStore,
+	processor WorkerTaskProcessor, logger log.Logger,
+) TaskQueue {
+	qCfg := cfg.AsyncService.WorkerTaskQueue
+
+	return &workerTaskQueueSQLImpl{
+		shardId: shardId,
+		store:   store,
+		logger:  logger.WithTags(tag.Shard(shardId)),
+		rootCtx: rootCtx,
+		cfg:     cfg,
+
+		pollTimer:                 NewLocalTimerGate(logger),
+		commitTimer:               NewLocalTimerGate(logger),
+		processor:                 processor,
+		tasksToCommitChan:         make(chan persistence.WorkerTask, qCfg.ProcessorBufferSize),
+		currentReadCursor:         0,
+		pendingTaskSequenceToPage: make(map[int64]*workerTaskPage),
+	}
+}
+
+func (w *workerTaskQueueSQLImpl) Stop(ctx context.Context) error {
+	// a final attempt to commit the completed page
+	return w.commitCompletedPages(ctx)
```

Comment:
The processed task in incompleted pages will be executed again when starting the server next time?

Reply:
Yes that's expected. When it executes again, it will load the AsyncStateExecution and found that it's already executed, then it will skip.